### PR TITLE
Fix #19030: Add DI container usage to Widget end()

### LINKED ISSUE #19030: base\Widget end() method throws Invalid Call exception if i override widget class in definition
### What steps will reproduce the problem?
Add Pjax to definitions

    \yii\widgets\Pjax::class => [
        'class' => \system\widgets\Pjax::class
    ]

### What is the expected result?
\system\widgets\Pjax::class class will be used insted of \yii\widgets\Pjax::class

### What do you get instead?
Expecting end() of system\widgets\Pjax, found yii\widgets\Pjax

### Additional info

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.43
| PHP version      | 7.4.3
| Operating system | Windows

### Proposal:

    public static function end() 
    {

        if (!empty(self::$stack)) {
            $widget = array_pop(self::$stack);

            $calledClass = get_called_class();
            if(Yii::$container->has($calledClass) && isset(Yii::$container->getDefinitions()[$calledClass]['class'])) {
                $calledClass = Yii::$container->getDefinitions()[$calledClass]['class'];
            }

            if (get_class($widget) === $calledClass) {
                /* @var $widget Widget */
                if ($widget->beforeRun()) {
                    $result = $widget->run();
                    $result = $widget->afterRun($result);
                    echo $result;
                }

                return $widget;
            }

            throw new InvalidCallException('Expecting end() of ' . get_class($widget) . ', found ' . get_called_class());
        }

        throw new InvalidCallException('Unexpected ' . get_called_class() . '::end() call. A matching begin() is not found.');
    }

### PR DESCRIPTION
**Pull request by papppeter**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | #19030

### CODE REVIEW

File: tests/framework/base/WidgetTest.php [hunk 1]
```diff
@@ -46,6 +48,27 @@ public function testBeginEnd()
         $this->assertSame('<run-test>', $output);
     }
 
+    public function testDI()
```
Comment (WinterSilence) [inline tests/framework/base/WidgetTest.php:51]: Add comment
~~~php
/**
 * @covers Widget::begin
 * @covers Widget::end
 */
~~~
https://phpunit.readthedocs.io/en/9.5/annotations.html#covers

File: framework/base/Widget.php [hunk 2]
```diff
@@ -104,7 +104,13 @@ public static function end()
     {
         if (!empty(self::$stack)) {
             $widget = array_pop(self::$stack);
-            if (get_class($widget) === get_called_class()) {
+
+            $calledClass = get_called_class();
+            if(Yii::$container->has($calledClass) && isset(Yii::$container->getDefinitions()[$calledClass]['class'])) {
```
Comment (WinterSilence) [inline framework/base/Widget.php:109]: missed white space: `if (`

Comment (papppeter) [reply framework/base/Widget.php]: thanks for spotting, added the white space

File: tests/framework/base/WidgetTest.php (hunk 1 above)
Comment (papppeter) [reply tests/framework/base/WidgetTest.php:56]: Thanks for suggestion i added comments

**Comment by bizley** [comment]
Hm, maybe just drop the `@covers` - not really needed and looks like it's messing the tests

**Comment by papppeter** [comment]
> @bizley test method not match by pattern `test{$method}`, `@covers` define tested method(s)

can you help me what should i change? i am not that familiar with test docs

mybe test method need to change to testDIBeginEnd

**Comment by bizley** [comment]
Not really sure about what WinterSilence said but I'm pretty sure that `@covers` is not necessary. First remove it, and if it's still failing maybe change the name of the test from `testDI` to something longer like `testDependencyInjection`.

**Comment by papppeter** [comment]
> Not really sure about what WinterSilence said but I'm pretty sure that `@covers` is not necessary. First remove it, and if it's still failing maybe change the name of the test from `testDI` to something longer like `testDependencyInjection`.

i did it. i hope now it will be ok.

**Comment by WinterSilence** [comment]
@bizley it's work in any case, tags need to [Code Coverage Analysis](https://phpunit.readthedocs.io/en/9.5/code-coverage-analysis.html)

**Comment by WinterSilence** [comment]
@papppeter if your test method not match by pattern `testMethodName()`, then you can define covered method by `@covers`, see  [phpunit manual](https://phpunit.readthedocs.io/en/9.5/code-coverage-analysis.html#code-coverage-analysis-specifying-covered-parts)

**Comment by bizley** [comment]
No, tags are not needed for coverage to work, it's just a pointer for the user.

**Comment by papppeter** [comment]
I am happy with both solution. But @covers failed if i see it right without it all checks are passing. maybe i did not set @covers in the right way. I accept any solution you suggest.

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
Thanks.

**Pull request merged**